_posts/2024-03-20-strimzi-kraft-migration.md

```diff
@@ -0,0 +1,260 @@
+```
+
+The annotation change triggers the Strimzi operator to run the following steps:
+
+* deploy the KRaft controllers, which are configured with all the connection details to ZooKeeper and with the migration flag enabled;
+* configure the running brokers with the connection details to the KRaft controller quorum, the migration flag enabled and roll them;
+* on each reconciliation, check the status of the migration by verifying the `kafka.controller:type=KafkaController,name=ZkMigrationState` metric and update the `Kafka` custom resource to provide such status to the user through the `status.kafkaMetadataState` field;
+* when the migration is done, reconfigure brokers to be not connected to ZooKeeper anymore, with the migration flag disabled, and roll them again;
+
+As you can see, just applying one single annotation value has covered most of the phases in the migration procedure with the user doing actually nothing but following what the operator is running for them.
+
+While the migration is going on, you can see the cluster changing the metadata state during the migration process from `ZooKeeper` to `KRaftPostMigration`, and then waiting for the user to finalize it.
+
+```shell
+NAME         DESIRED KAFKA REPLICAS   DESIRED ZK REPLICAS   READY   METADATA STATE   WARNINGS
+my-cluster   3                        3                     True    ZooKeeper        
+my-cluster   3                        3                     True    ZooKeeper        
+my-cluster   3                        3                     True    ZooKeeper        True
+my-cluster   3                        3                             ZooKeeper        True
+my-cluster   3                        3                             ZooKeeper        True
+my-cluster   3                        3                             ZooKeeper        
+my-cluster   3                        3                             ZooKeeper        
+my-cluster   3                        3                             ZooKeeper        
+my-cluster   3                        3                     True    KRaftMigration   
+my-cluster   3                        3                     True    KRaftDualWriting   
+my-cluster   3                        3                     True    KRaftPostMigration   
+my-cluster   3                        3                     True    KRaftPostMigration   True
+```
+
+In this status, the cluster is still working in "dual-write" mode and the user can validate that everything is working fine before deciding to finalize the migration or rolling back to use ZooKeeper.
+If the `Kafka` custom resource has the `inter.broker.protocol.version` and `log.message.format.version` parameters set in the `spec.kafka.config` section, and because they are not supported in KRaft, the operator reports warnings into the status (see `WARNINGS` column).
+The user has to remove the parameters at the end of the process causing an additional rolling of the nodes.
```

Comment:
I'm thinking if it would make sense to have something like this:
```
Once the metadata state is in `KRaftPostMigration` state, you have two options how you can continue:

* finish the migration -- link to the "Migration finalization" section
* roll back to ZooKeeper -- link to the "Rollback support" section
```

I think it will make it more clear to the reader that from this state there are two options. Because if you read it as it is, this sentence can make it a bit confusing `When the migration is not finalized yet, so when the KRaft controllers are still connected to ZooKeeper, it is still possible to rollback.` -> "what is the exact moment I can still roll back to ZK?"

And yeah I know that it will "duplicate" the `deciding to finalize the migration or rolling back to use ZooKeeper.` part of the sentence.

Reply:
Good idea!